### PR TITLE
Deploy mongodb replicaset using petsets.

### PR DESCRIPTION
**Pull request by foxish**
https://github.com/kubernetes/kubernetes/issues/30289

Initial support for mongodb replicasets.

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/contrib/1512)

### CODE REVIEW

**Comment by foxish** [comment]
@bprashanth, PTAL and let me know if this is along the right lines. I'll be updating the README in line with the other examples.

File: pets/mongodb/init/on-start.sh [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        continue
+    fi
+    PEERS=("${PEERS[@]}" $LINE)
+done
+
+apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv 7F0CEB10 \
+ && echo 'deb http://downloads-distro.mongodb.org/repo/ubuntu-upstart dist 10gen' > /etc/apt/sources.list.d/mongodb.list \
+ && apt-get update \
+ && DEBIAN_FRONTEND=noninteractive apt-get install -y mongodb-org-server mongodb-org-shell \
+ && rm -rf /var/lib/apt/lists/*
```
Comment (bprashanth) [inline pets/mongodb/init/on-start.sh]: If theres an existing mongo image you can simply use it as the base. I might've done this in places that don't have a docker image, but it's a pain to continuously do this for every pod. I ended up baking this into the image we use in e2es for redis and mysql: https://github.com/kubernetes/kubernetes/tree/master/test/images/pets/redis. 

It's preferable to not use a custom image in the app contaienr, inita containers less so.

File: pets/mongodb/mongodb.yaml [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+  template:
+    metadata:
+      labels:
+        app: mongodb
+      annotations:
+        pod.alpha.kubernetes.io/initialized: "true"
+        pod.alpha.kubernetes.io/init-containers: '[
+            {
+                "name": "install",
+                "image": "docker.io/foxish/mongodb-install:0.1",
```
Comment (bprashanth) [inline pets/mongodb/mongodb.yaml]: if you're using a custom image anyway no need to download per pod

File: pets/mongodb/mongodb.yaml [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+            - -c
+            - "/usr/bin/mongo --eval 'printjson(db.serverStatus())'"
+          initialDelaySeconds: 5
+          timeoutSeconds: 5
+        volumeMounts:
+        - name: datadir
+          mountPath: /data/db
+        - name: config
+          mountPath: /etc/mongodb
+        - name: workdir
```
Comment (bprashanth) [inline pets/mongodb/mongodb.yaml]: why does this need workdir? config and data i get

File: pets/mongodb/init/on-start.sh (hunk 1 above)
Comment (foxish) [inline pets/mongodb/init/on-start.sh]: Fixed, switched to mongo:3.2 official image.

File: pets/mongodb/mongodb.yaml (hunk 2 above)
Comment (foxish) [inline pets/mongodb/mongodb.yaml]: Done.

File: pets/mongodb/mongodb.yaml (hunk 3 above)
Comment (foxish) [inline pets/mongodb/mongodb.yaml]: Removed. Thanks!

**Comment by foxish** [comment]
@bprashanth Fixed the issues. PTAL.

File: pets/mongodb/README.md [hunk 4]
```diff
@@ -0,0 +1,4 @@
+# TODO
```
Comment (ingvagabund) [inline pets/mongodb/README.md]: Can you write down some steps how to setup kubernetes environment (if necessary) and how to run ``mongodb`` petset? Something similar to [1].

[1] https://github.com/kubernetes/contrib/blob/master/pets/zookeeper/README.md

File: pets/mongodb/mongodb.yaml [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+  volumeClaimTemplates:
+  - metadata:
+      name: datadir
+      annotations:
+        volume.alpha.kubernetes.io/storage-class: anything
+    spec:
+      accessModes: [ "ReadWriteOnce" ]
+      resources:
+        requests:
+          storage: 10Gi
```
Comment (ingvagabund) [inline pets/mongodb/mongodb.yaml]: Is there a recommended size for datadir?

File: pets/mongodb/README.md (hunk 4 above)
Comment (foxish) [inline pets/mongodb/README.md]: Done

File: pets/mongodb/mongodb.yaml (hunk 5 above)
Comment (foxish) [inline pets/mongodb/mongodb.yaml]: I couldn't find one. I think it preallocates 2-3 GB at first, based on what I read, but there wasn't really a recommended size specified.

**Comment by sunruh** [comment]
@foxish I think logging should not be configured in the config file to not change the behavior of the "standard" mongodb image, as setting the logpath back to stdout via the command line is not possible, AFAIK.

File: pets/mongodb/README.md [hunk 6]
```diff
@@ -0,0 +1,146 @@
+# Zookeeper
```
Comment (ingvagabund) [inline pets/mongodb/README.md]: s/Zookeeper/MongoDB

Comment (foxish) [inline pets/mongodb/README.md]: Oops! Fixed... Thanks :)

**Comment by foxish** [comment]
@sunruh That is a good point. Thanks for pointing it out. 
I've removed logging from the config file. It should be possible now to override it.

**Comment by eyalse** [comment]
@foxish im gettin a 
`error validating "mongodb.yaml": error validating data: the server could not find the requested resource (get .apps);`
on 1.3.5

any idea?

**Comment by foxish** [comment]
@eyalse Could you do `kubectl version` and tell me your client and server versions? Are they both 1.3.5?

**Comment by ingvagabund** [comment]
@eyalse did you updated to newer version of kubernetes or did you installed the kubernetes on fresh system? If it is the former, try to remove ``~/.kube/schema``.

**Comment by eyalse** [comment]
@foxish & @ingvagabund heya
im deploying a fresh cluster via gke.
`kubectl version` returns:
```
Client Version: version.Info{Major:"1", Minor:"3", GitVersion:"v1.3.3", GitCommit:"c6411395e09da356c608896d3d9725acab821418", GitTreeState:"clean", BuildDate:"2016-07-22T20:29:38Z", GoVersion:"go1.6.2", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"3", GitVersion:"v1.3.5", GitCommit:"b0deb2eb8f4037421077f77cb163dbb4c0a2a9f5", GitTreeState:"clean", BuildDate:"2016-08-11T20:21:58Z", GoVersion:"go1.6.2", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by eyalse** [comment]
i think that i need to enable the `apiVersion: "apps/v1alpha1"` somehow.

**Comment by bprashanth** [comment]
maybe we should migrate this to charts, https://github.com/kubernetes/charts/pull/38 has set precedence

**Comment by viglesiasce** [comment]
Yes! Please bring this on over to kubernetes/charts repository. In the incubator directory for now until PetSet is out of alpha.

**Pull request closed without merging**